File: js/src/store.js
/**
 * JSON persistence backend (SPC section 22): a single canonical JSON file
 * at `review.store` (default `.tux/feedback.json`), plus review sessions
 * under `.tux/sessions/` and incorporation batches under
 * `.tux/incorporations/`. The logical schema is independent of the
 * storage backend; agents go through the CLI/API, never the file layout.
 *
 * @module store
 */
import { existsSync, mkdirSync, readFileSync, readdirSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';
import { canonicalJson } from './canonical.js';
import { SCHEMA_VERSION } from './schema.js';

export function storeFilePath(cwd, config) {
  return join(cwd, config.review.store);
}

export function loadStore(cwd, config) {
  const p = storeFilePath(cwd, config);
  if (!existsSync(p)) return { schema_version: SCHEMA_VERSION, project_id: config.project_id, feedback: [] };
  const store = JSON.parse(readFileSync(p, 'utf8'));
  if (store.feedback === undefined) store.feedback = [];
  return store;
}

export function saveStore(cwd, config, store) {
  const p = storeFilePath(cwd, config);
  mkdirSync(join(cwd, '.tux'), { recursive: true });
  writeFileSync(p, canonicalJson(store) + '\n', 'utf8');
}

export function storeExists(cwd, config) {
  return existsSync(storeFilePath(cwd, config));
}

/** Sessions directory handling. */
export function sessionsDir(cwd) {
  return join(cwd, '.tux', 'sessions');
}

export function ensureSession(cwd, config, sessionId, environment, timestamp) {
  mkdirSync(sessionsDir(cwd), { recursive: true });
  const p = join(sessionsDir(cwd), `${sessionId}.json`);
  if (existsSync(p)) return JSON.parse(readFileSync(p, 'utf8'));
  const session = {
    session_id: sessionId,
    project_id: config.project_id,
    environment,
    status: 'active',
    created_at: timestamp,
  };
  writeFileSync(p, canonicalJson(session) + '\n', 'utf8');
  return session;
}

/** Incorporation batches directory handling. */
export function incorporationsDir(cwd) {
  return join(cwd, '.tux', 'incorporations');
}

export function countIncorporationBatches(cwd) {
  const dir = incorporationsDir(cwd);
  return existsSync(dir) ? readdirSync(dir).filter((f) => f.endsWith('.json')).length : 0;
}

export function saveIncorporationBatch(cwd, batch) {
  mkdirSync(incorporationsDir(cwd), { recursive: true });
  const p = join(incorporationsDir(cwd), `${batch.batch_id}.json`);
  writeFileSync(p, canonicalJson(batch) + '\n', 'utf8');
  return p;
}

File: py/tux/store.py
"""JSON persistence backend (SPC section 22): a single canonical JSON
file at ``review.store`` (default ``.tux/feedback.json``), plus review
sessions under ``.tux/sessions/`` and incorporation batches under
``.tux/incorporations/``."""
from __future__ import annotations

import json
from pathlib import Path

from .canonical import canonical_json
from .schema import SCHEMA_VERSION


def store_file_path(cwd: str, config: dict) -> Path:
    return Path(cwd) / config["review"]["store"]


def load_store(cwd: str, config: dict) -> dict:
    p = store_file_path(cwd, config)
    if not p.exists():
        return {"schema_version": SCHEMA_VERSION, "project_id": config["project_id"], "feedback": []}
    store = json.loads(p.read_text(encoding="utf-8"))
    if "feedback" not in store:
        store["feedback"] = []
    return store


def save_store(cwd: str, config: dict, store: dict) -> None:
    p = store_file_path(cwd, config)
    p.parent.mkdir(parents=True, exist_ok=True)
    p.write_text(canonical_json(store) + "\n", encoding="utf-8", newline="\n")


def store_exists(cwd: str, config: dict) -> bool:
    return store_file_path(cwd, config).exists()


def sessions_dir(cwd: str) -> Path:
    return Path(cwd) / ".tux" / "sessions"


def ensure_session(cwd: str, config: dict, session_id: str, environment: str, timestamp: str) -> dict:
    d = sessions_dir(cwd)
    d.mkdir(parents=True, exist_ok=True)
    p = d / f"{session_id}.json"
    if p.exists():
        return json.loads(p.read_text(encoding="utf-8"))
    session = {
        "session_id": session_id,
        "project_id": config["project_id"],
        "environment": environment,
        "status": "active",
        "created_at": timestamp,
    }
    p.write_text(canonical_json(session) + "\n", encoding="utf-8", newline="\n")
    return session


def incorporations_dir(cwd: str) -> Path:
    return Path(cwd) / ".tux" / "incorporations"


def count_incorporation_batches(cwd: str) -> int:
    d = incorporations_dir(cwd)
    return len([f for f in d.iterdir() if f.name.endswith(".json")]) if d.exists() else 0


def save_incorporation_batch(cwd: str, batch: dict) -> Path:
    d = incorporations_dir(cwd)
    d.mkdir(parents=True, exist_ok=True)
    p = d / f"{batch['batch_id']}.json"
    p.write_text(canonical_json(batch) + "\n", encoding="utf-8", newline="\n")
    return p
